8297427: Avoid keeping class loaders alive when executing ClassLoaderStatsVMOperation

kstefanj opened the pull request:
Please review this change to avoid keeping classes alive only due to the `ClassLoaderStatsVMOperation`.

**Summary**
The `ClassLoaderStatsVMOperation` is gathering statistics about the active class loaders in a safepoint. The way the `ClassLoaderDataGraph` is iterated will keep the class loaders live. This is not really needed since everything is done in a safepoint and nothing needs to be explicitly kept alive. This has not been a problem prior to concurrent class unloading in ZGC. With fully concurrent class unloading a `ClassLoaderStatsVMOperation` can occur during a collection and more classes than needed might be kept alive. This could in turn lead to premature Metaspace OOM. 

The solution is to not keep the class loaders alive due to the iteration in `ClassLoaderStatsVMOperation`.

**Testing**
* Added a new test that covers the two different ways a class could previously be kept alive by the VM operation. The test passes after the fix but failed before.  
* Mach5 tier 1-3

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8297427](https://bugs.openjdk.org/browse/JDK-8297427): Avoid keeping class loaders alive when executing ClassLoaderStatsVMOperation


### Reviewers
 * [Erik Österlund](https://openjdk.org/census#eosterlund) (@fisk - **Reviewer**) ⚠️ Review applies to [424e0f9d](https://git.openjdk.org/jdk/pull/11300/files/424e0f9d831279ba2d1986ebacb499f8d4a6c078)
 * [Stefan Karlsson](https://openjdk.org/census#stefank) (@stefank - **Reviewer**) ⚠️ Review applies to [f6dba7f2](https://git.openjdk.org/jdk/pull/11300/files/f6dba7f200da4311399157760a97e5b481985abc)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11300/head:pull/11300` \
`$ git checkout pull/11300`

Update a local copy of the PR: \
`$ git checkout pull/11300` \
`$ git pull https://git.openjdk.org/jdk pull/11300/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11300`

View PR using the GUI difftool: \
`$ git pr show -t 11300`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11300.diff">https://git.openjdk.org/jdk/pull/11300.diff</a>

</details>

fisk reviewed (approved):
Looks good to me. Nice test!

kstefanj commented:
Reverted the use of templates after some internal discussions. Also added a print of the object in the test to ensure the class is kept alive even if `-Xcomp` is used and ergonomically GCs are triggered before the one invoked by the test.

stefank reviewed (approved):
Looks good!

dholmes-ora commented on test/hotspot/jtreg/runtime/ClassUnload/UnloadTestDuringClassLoaderStatsVMOperation.java:
```diff
@@ -0,0 +1,149 @@
+ */
+
+/*
+ * @test Class unloading test while triggering execution of ClassLoaderStats VM operations
+ * @summary Make sure class unloading occur even if ClassLoaderStats VM operations are executed
```
Having a comment on `@test` and providing `@summary` seems redundant. I would just leave the summary.

You could/should(?) add a `@bug` line too.

dholmes-ora commented on test/hotspot/jtreg/runtime/ClassUnload/UnloadTestDuringClassLoaderStatsVMOperation.java:
```diff
@@ -0,0 +1,149 @@
+ * @summary Make sure class unloading occur even if ClassLoaderStats VM operations are executed
+ * @requires vm.opt.final.ClassUnloading
+ * @modules java.base/jdk.internal.misc
+ * @library /test/lib
+ * @library classes
```
You can put multiple libraries on one `@library` - could save some jtreg  parsing time.

dholmes-ora commented on test/hotspot/jtreg/runtime/ClassUnload/UnloadTestDuringClassLoaderStatsVMOperation.java:
```diff
@@ -0,0 +1,149 @@
+ * @run driver jdk.test.lib.helpers.ClassFileInstaller jdk.test.whitebox.WhiteBox
+ * @run main/othervm -Xbootclasspath/a:. -XX:+UnlockDiagnosticVMOptions -XX:+WhiteBoxAPI -Xlog:gc*,class+unload=debug UnloadTestDuringClassLoaderStatsVMOperation
+ */
+import jdk.test.whitebox.WhiteBox;
+
+import java.net.URLClassLoader;
+
+import jdk.test.lib.classloader.ClassUnloadCommon;
```
Should imports be sorted and grouped ie.
```
import java.net.URLClassLoader;

import jdk.test.lib.classloader.ClassUnloadCommon;
import jdk.test.whitebox.WhiteBox;
```
? Not sure if we have a style-guide for this.

dholmes-ora commented on test/hotspot/jtreg/runtime/ClassUnload/UnloadTestDuringClassLoaderStatsVMOperation.java:
```diff
@@ -0,0 +1,149 @@
+
+    public static void main(String args[]) throws Exception {
+        // Create a thread forcing ClassLoaderStats VM operations.
+        Runnable task = () -> {
+            while (true) {
+                wb.forceClassLoaderStatsSafepoint();
+            }
+        };
+        var clsThread = new Thread(task);
```
You don't need `task` here as you can just write the above as:
```
var clsThread = new Thread(() -> {
            while (true) {
                wb.forceClassLoaderStatsSafepoint();
            }
        });
```

dholmes-ora commented on test/hotspot/jtreg/runtime/ClassUnload/UnloadTestDuringClassLoaderStatsVMOperation.java:
```diff
@@ -0,0 +1,149 @@
+
+        ClassUnloadCommon.failIf(!wb.isClassAlive(className), className + " should be loaded and live");
+
+        // Printing the object to ensure the class is kept alive. If the test
```
We should be using `Reference.reachabilityFence` to keep objects alive.

dholmes-ora reviewed (commented):
A few drive-by comments on the test.

kstefanj commented:
Thanks for your comments @dholmes-ora and for your reviews @stefank and @fisk.
/integrate
